workflow-ui/backstack-android/src/main/java/com/squareup/workflow1/ui/backstack/StateRegistryHolder.kt

```diff
@@ -0,0 +1,222 @@
+   * [saveRegistryController].
+   *
+   * If this object has not been restored yet, this method will start listening to the parent
+   * lifecycle to know when to restore.
+   *
+   * Must be accompanied by a call to [detachFromParentRegistry] when the view is detached.
+   */
+  fun attachToParentRegistry(
+    key: String,
+    parentOwner: SavedStateRegistryOwner
+  ) {
+    check(parentRegistryOwner == null) {
+      "Expected StateRegistryHolder to be detached before re-attaching. " +
+        "Already attached to $parentRegistryOwner but trying to attach to $parentOwner"
+    }
+    this.parentRegistryOwner = parentOwner
+    this.parentKey = key
+
+    // We can only be restored once.
+    if (isRestored) return
+
+    val parentRegistry = parentOwner.savedStateRegistry
+    val parentLifecycle = parentOwner.lifecycle
+
+    // If the key is already registered, SavedStateRegistry throws an exception that doesn't provide
+    // enough information to troubleshoot, so we add some ourselves.
+    try {
+      parentRegistry.registerSavedStateProvider(key, ::saveToBundle)
+    } catch (e: IllegalArgumentException) {
+      throw IllegalArgumentException(
+        "Error registering StateRegistryHolder as SavedStateProvider with key \"$key\" on parent " +
+          "SavedStateRegistryOwner $parentOwner.\n" +
+          "You might need to set a view ID on your BackStackContainer or wrap your " +
+          "BackStackScreen with a Named rendering.",
+        e
+      )
+    }
+
+    // Even if the parent lifecycle is in a state further than CREATED, new observers are sent all
+    // the lifecycle events required to catch the observer up to the current state, so this handles
+    // both the cases where we're ready to immediately restore, and where we have to wait.
+    parentLifecycle.addObserver(lifecycleObserver)
+  }
+
+  /**
+   * Must be called when the owning view detaches from the window.
+   *
+   * Stops listening to the parent lifecycle and unregisters from the parent registry.
+   */
+  fun detachFromParentRegistry() {
+    val parentOwner = checkNotNull(parentRegistryOwner) {
+      "Expected to be detached only when attached."
+    }
+    parentOwner.savedStateRegistry.unregisterSavedStateProvider(parentKey!!)
+    parentOwner.lifecycle.removeObserver(lifecycleObserver)
+    parentRegistryOwner = null
+    parentKey = null
+  }
+
+  /**
+   * Asks [controller] to [save itself][SavedStateRegistryController.performSave] and caches that
+   * saved state for when our parent registry gets saved. This should be called any time the view
+   * associated with the child controller is going away but may be shown again later, and from the
+   * [onWillSave] callback.
+   *
+   * @param key The key used to distinguish [controller] from other controllers saved to this
+   * object. Must be the same key used to [restore][restoreRegistryControllerIfReady] the controller
+   * later.
+   */
+  fun saveRegistryController(
```

Comment:
I figured it wasn't necessary to be quite that explicit, since the idea of saving the controller itself doesn't make any sense, and the entire point of these controllers is to save their state.